src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,758 @@
+
+                                try
+                                {
+                                    var itemType = binder.GetStreamItemType(id);
+                                    if (reader.TokenType != JsonTokenType.Null)
+                                    {
+                                        using var token = JsonDocument.ParseValue(ref reader);
+                                        item = BindType(token.RootElement, itemType);
+                                    }
+                                }
+                                catch (Exception ex)
+                                {
+                                    return new StreamBindingFailureMessage(id, ExceptionDispatchInfo.Capture(ex));
+                                }
+                            }
+                            else if (reader.TextEquals(ArgumentsPropertyNameBytes))
+                            {
+                                reader.CheckRead();
+
+                                int initialDepth = reader.CurrentDepth;
+                                if (reader.TokenType != JsonTokenType.StartArray)
+                                {
+                                    throw new InvalidDataException($"Expected '{ArgumentsPropertyName}' to be of type {SystemTextJsonExtensions.GetTokenString(JsonTokenType.StartArray)}.");
+                                }
+
+                                hasArguments = true;
+
+                                if (string.IsNullOrEmpty(target))
+                                {
+                                    // We don't know the method name yet so just store the array in JsonDocument
+                                    argumentsToken = JsonDocument.ParseValue(ref reader);
+                                }
+                                else
+                                {
+                                    try
+                                    {
+                                        var paramTypes = binder.GetParameterTypes(target);
+                                        if (reader.TokenType != JsonTokenType.Null)
+                                        {
+                                            using var token = JsonDocument.ParseValue(ref reader);
+                                            arguments = BindTypes(token.RootElement, paramTypes);
+                                        }
+                                    }
+                                    catch (Exception ex)
+                                    {
+                                        argumentBindingException = ExceptionDispatchInfo.Capture(ex);
+
+                                        // Could be at any point in argument array JSON when an error is thrown
+                                        // Read until the end of the argument JSON array
+                                        while (reader.CurrentDepth == initialDepth && reader.TokenType == JsonTokenType.StartArray ||
```

Comment:
Is there good unit tests coverage here? I recall @ahsonkhan discussing changing the reader depth logic. I don't know whether that has happened or not yet.

Reply:
We have the ArgumentBindingFailure tests, but because we're using JsonDocument.Parse right now, this code path isn't actually used. It will be used once we move off JsonDocument.

Reply:
> I recall @ahsonkhan discussing changing the reader depth logic. I don't know whether that has happened or not yet.

It did happen. It matches Json.NET semantics now.

See: https://github.com/dotnet/corefx/pull/36116